File: 11-container-with-most-water/solution.ts
export function maxArea(height: number[]): number {
  let i = 0;
  let j = height.length - 1;
  let maxArea = 0;
  while (i < j) {
    const heightI = height[i] as number;
    const heightJ = height[j] as number;
    const area = (j - i) * Math.min(heightI, heightJ);
    maxArea = Math.max(maxArea, area);

    if (heightI < heightJ) {
      ++i;
    } else if (heightI > heightJ) {
      --j;
    } else {
      ++i;
      --j;
    }
  }

  return maxArea;
}

File: 11-container-with-most-water/solution.cpp
#include <vector>

class Solution
{
public:
  int maxArea(std::vector<int> &height)
  {
    size_t i = 0;
    size_t j = height.size() - 1;
    int max_area = 0;
    while (i < j)
    {
      int area = (j - i) * std::min(height[i], height[j]);
      max_area = std::max(max_area, area);

      if (height[i] < height[j])
      {
        ++i;
      }
      else if (height[i] > height[j])
      {
        --j;
      }
      else
      {
        ++i;
        --j;
      }
    }

    return max_area;
  }
};